### ISSUE TITLE
vnd with vlan over aggr in promisc see all traffic

### ISSUE THREAD

**Issue by sjorge** [joyent/smartos-live#716]
Sorry for not providing any meaning full packet or log capture, as I had to ditch the setup and fall back my old setup due to it causing a bit of havoc.

It seems a vnic (well in this case a vnd nic as it is KVM) set with ```allow_unfiltered_promisc``` to ```true``` will see traffic for all vlans if is placed over an aggr that uses tagged vlans.

I have a similar aggr on 2 cns that have tagged vlans 10,20,30 and 40.
```
trunk0_aggr=0c:c4:7a:32:27:84,0c:c4:7a:32:27:85,0c:c4:7a:32:27:86,0c:c4:7a:32:27:87
trunk0_lacp_mode=active
trunk0_mtu=9000
trunk_nic=trunk0
```
On each CN I was running a opnsense kvm guest in a HA cluster. This uses CARP to figure out who gets the IP.

I also have had 2 zones for a long time using ucarp in base64 and those work fine with this config:
```
{
  "interface": "net0",
  "mac": "xxx:xx:xx:xx:05:01",
  "vlan_id": 5,
  "nic_tag": "external",
  "netmask": "255.255.255.0",
  "ip": "192.168.0.22",
  "ips": [
    "192.168.0.22/24"
  ],
  "allowed_ips": [
    "192.168.0.20/32",
    "224.0.0.18/32"
  ],
  "allow_mac_spoofing": true,
  "mtu": 1500
}
```

The VIP comes online and is usable.

Using the same setup for the opnsense KVM boxes does not seem to work. They agree on who the master is but no traffic for/from the VIP working, using a nic in unfiltered_promisc makes the VIP work.

```
{
  "interface": "net0",
  "mac": "xx:xx:xx:ff:40:01",
  "vlan_id": 40,
  "nic_tag": "trunk",
  "ip": "dhcp",
  "ips": [
    "172.16.40.11/24"
  ],
  "model": "virtio",
  "allow_dhcp_spoofing": true,
  "allow_mac_spoofing": true,
  "allow_unfiltered_promisc": true,
  "mtu": 1500
}
```

All things are fine with just one vnd nic, once I add the other 3 (4 in total). Things never get stable.
Each interface in the kvm seems to see traffic for all vlans, at least the multicast traffic for CARP.

Resulting in a unstable systems that is not usable. 

The behavior seems seeing all traffic for the vlan which the vnd nic belongs too is what I would expect, not seeing all traffic for all vlan on the under laying aggr. (Note it may actually only be broadcast and multicast traffic, which should still not leak between vlans!) I never managed to get more logs because the kvm would become unresponsive rather quickly :(

**Comment by rmustacc** [comment]
The point of allow_unfiltered_promisc is to see all the traffic that the underlying device sees. In general, it's not an option that you want to be using for reasons such as this.

**Comment by rmustacc** [comment]
It'd probably be more helpful to understand what you're trying to do that led you down the path of wanting unfiltered promiscuous mode to be enabled. You should be able to get all broadcast and unicast traffic by default and could add a secondary mac alternatively if there's something else. I'm not sure if the carp metadata traffic and heartbeating was multicast, but I thought some amount of it was.

**Comment by sjorge** [comment]
@rmustacc rebuilding my gateway slowly, my old one got corrupt.
I was trying to setup the new one HA. Once I have everything running smoothly I'll get back to you with some more details.

**Comment by plitc** [comment]
We currently have the same problem

> OpnSense A (CARP MASTER) <-> SmartOS 1 <=== (LACP) ===> Switch <=== (LACP) ===> SmartOS 2 <-> OpnSense B (CARP SLAVE)

SmartOS Config:

> admin_nic=XX:XX:XX:XX:XX:08
admin_ip=XXX.XXX.XXX.16
admin_netmask=255.255.255.0
admin_network=
admin_gateway=XXX.XXX.XXX.16

> aggr1_aggr=XX:XX:XX:XX:XX:09,XX:XX:XX:XX:XX:0a,XX:XX:XX:XX:XX:0b
aggr1_lacp_mode=active
vlan_nic=aggr1

> vlan0_ip=10.0.0.16
vlan0_netmask=255.255.252.0
vlan0_gateway=10.0.0.250
vlan0_vlan_id=100

... and more ...

KVM - OpnSense .json

> {
      "__comment" : "current untagged subnet",
      "nic_tag": "admin",
      "ip": "XXX.XXX.XXX.243",
      "netmask": "255.255.255.0",
      "gateway": "XXX.XXX.XXX.27",
      "model": "virtio",
      "primary": true,
      "__comment" : "carp ips",
      "allowed_ips": [
         "XXX.XXX.XXX.246"
      ],
      "allow_dhcp_spoofing": true,
      "allow_ip_spoofing": true,
      "allow_mac_spoofing": true,
      "allow_restricted_traffic": true,
      "allow_unfiltered_promisc": true
    },

... and 6 vlan interfaces

![smartos_kvm_opnsense_carp_1](https://cloud.githubusercontent.com/assets/6014760/25649793/c1b57ed2-2fd8-11e7-9951-33aad6dfadb5.jpg)

in Grafana we see very high interrupts

- If we enable (opnsense) netflow, the cpus 100% is loaded

without:
> "allow_unfiltered_promisc": true

CARP does not work

TCPdump Example

> 08:28:47.960406 00:00:5e:00:01:07 > 01:00:5e:00:00:12, ethertype IPv4 (0x0800), length 70: 10.0.20.254 > 224.0.0.18: VRRPv2, Advertisement, vrid 7, prio 0, authtype none, intvl 1s, length 36

**Comment by sjorge** [comment]
That is almost exactly my setup, I also tried with pfsense with the same result.

I have 5 vlans going over the LACP aggr. The *sense box has a vnd nic in 4 of the 5 vlans.
Each VLAN now seems to see all traffic for all vlans, *sense now gets confused on which port which VIP belongs too. Without promisc VIP get negotiated fine but there is no traffic.

**Comment by plitc** [comment]
I think the **promiscuous mode** is always a bad idea (poor performance, maybe only for port mirror debugging purposes useful)

I am unclear why it does not work with:
> "allow_ip_spoofing": true,
"allow_mac_spoofing": true,

**Comment by sjorge** [comment]
Yes, but from the VM's perspective it has a virtio nic that gets untagged traffic.
The SmartOS host untags it, when it says hey nic go into promisc mode it gets all kinds of stuff it did not expect.

From what I remember on ESXi for example it will only see all traffic for the vlan the nic is on as untagged but not other traffic.

**Comment by plitc** [comment]
this error description makes sense

**Comment by plitc** [comment]
On another machine observed:

> OpnSense C (CARP SLAVE) <-> SmartOS 3 <=== (no LACP / but mixed untagged + vlan) ===> Switch

SmartOS Config:

> admin_nic=XX:XX:XX:XX:XX:43
admin_ip=XXX.XXX.XXX.128
admin_netmask=255.255.255.0
admin_network=
admin_gateway=XXX.XXX.XXX.128

> admin0_ip=10.0.0.128
admin0_netmask=255.255.252.0
admin0_gateway=10.0.0.250
admin0_vlan_id=100

> admin1_ip=10.0.4.128
admin1_netmask=255.255.252.0
admin1_gateway=10.0.4.250
admin1_vlan_id=101

... and more ...

tcpdump:

> tcpdump -e -n -i vtnet0 'not port 222'

**the same traffic flood**

> 10:09:18.836489 00:00:5e:00:01:08 > 01:00:5e:00:00:12, ethertype IPv4 (0x0800), length 70: 10.0.24.254 > 224.0.0.18: VRRPv2, Advertisement, vrid 8, prio 0, authtype none, intvl 1s, length 36

> 10:09:18.893257 00:00:5e:00:01:04 > 01:00:5e:00:00:12, ethertype IPv4 (0x0800), length 70: 10.0.8.254 > 224.0.0.18: VRRPv2, Advertisement, vrid 4, prio 0, authtype none, intvl 1s, length 36

> 10:09:19.131960 XX:XX:XX:XX:XX:8c > XX:XX:XX:XX:XX:0e, ethertype LLDP (0x88cc), length 180: LLDP, length 166: HP ProCurve Switch 4108GL

**Comment by plitc** [comment]
- http://conferences.sigcomm.org/sigcomm/2009/workshops/visa/papers/p53.pdf
- http://www.oracle.com/technetwork/articles/servers-storage-admin/sol11ecrossbow-186794.pdf

**maybe an internal crossbow issue?**

1. the physical interface get the "tagged" package
2. the zone has defined (in /etc/zones/UUID.xml):
`<network physical="net1" mac-addr="XX:XX:XX:XX:XX:9a" vlan-id="100" global-nic="admin">`
3. SmartOS HOST environment
> dladm show-vnic

`net1         bge0       0     XX:XX:XX:XX:XX:9a fixed       100  UUID`

4. Link Property

> dladm show-linkprop

`net1         autopush        rw   --             --             --`
`net1         zone            rw   UUID -- --`
`net1         state           r-   up             up             up,down`
`net1         mtu             rw   1500           1500           1500`
`net1         secondary-macs  rw   --             --             --`
`net1         maxbw           rw   --             --             --`
`net1         cpus            rw   --             --             --`
`net1         cpus-effective  r-   5-6            --             --`
`net1         pool            rw   --             --             --`
`net1         pool-effective  r-   --             --             --`
`net1         priority        rw   high           high           low,medium,high`
`net1         tagmode         rw   vlanonly       vlanonly       normal,vlanonly`
`net1         protection      rw   --             --             mac-nospoof,`
`                                                                restricted,`
`                                                                ip-nospoof,`
`                                                                dhcp-nospoof`
`net1         promisc-filtered rw  off            on             off,on`
`net1         allowed-ips     rw   --             --             --`
`net1         allowed-dhcp-cids rw --             --             --`
`net1         rxrings         rw   --             --             --`
`net1         rxrings-effective r- --             --             --`
`net1         txrings         rw   --             --             --`
`net1         txrings-effective r- --             --             --`
`net1         txrings-available r- 0              --             --`
`net1         rxrings-available r- 0              --             --`
`net1         rxhwclnt-available r- 0             --             --`
`net1         txhwclnt-available r- 0             --             --`

5. KVM Environment (in /zones/UUID/root/startvm)
`"-device" "virtio-net-pci,mac=XX:XX:XX:XX:XX:9a,tx=timer,x-txtimer=200000,x-txburst=128,vlan=1" "-net" "vnic,name=net1,vlan=1,ifname=net1,ip=10.0.0.253,netmask=255.255.252.0"`

The scheme is clearly defined

**Comment by sjorge** [comment]
I keep thinking of a switch vs hub scenario.

With promisc on the vnic acts like it is plugged into a hub (gets all traffic with tags in tact, ...)
While we want it to act like it is plugged into a promisc switch port.

We want it to get all traffic for the vlan it was assigned as untagged. While all other traffic is dropped for the vm.

**Comment by plitc** [comment]
on the **SmartOS 3** Machine runs a second KVM: **OpnSense D** (no CARP)

only with:

`<net-attr name="allow_ip_spoofing" value="true"/>`

option, so the tcpdump interface / vlan traffic flow works as expected

**conclusion**:
i think ... the promiscuous mode should show only vlan tag-specific data from the vnic ... not from the entire (smartos) host interface

**Comment by rmustacc** [comment]
This happens for a combination of reasons:

1.  We strip the tags for all traffic
2. In all cases you've explicitly asked for _all_ traffic from the underlying aggr to be passed to the VM rather than just the traffic that'd normally go there for the switch.

I can't stress enough that setting `allow_unfiltered_promisc` is generally not what you want to do, particularly in multi-VLAN scenarios. There's no surprise that you're seeing traffic from all the VLANs and potentially floods. This option changes what is and isn't filtered for a given VNIC. Without this option, the VNIC's promiscuous mode would behave like it's been plugged into a switch more unexpectedly.

If you only want traffic from the vnic, don't set `allow_unfiltered_promisc`.

**Comment by sjorge** [comment]
So the question is then, how can we allow the nic to get all traffic (untagged) from the specified vlan. Without getting all other traffic? I assume we just can't at the moment?

**Comment by rmustacc** [comment]
Correct, there's no per-VLAN notion of this at this time. Can you better describe why you want all the traffic from the VLAN promiscuously? Is there a reason that unicast for the VNIC + broadcast/multicast isn't sufficient? Do you need all unicast MACs on the VLAN or just a few?

**Comment by sjorge** [comment]
Same reason as @plitc we want opnsense/pfsense to work in HA mode.
For some reason as mentioned above "allow_ip_spoofing": true, "allow_mac_spoofing": true,
and optionally "allow_restricted_traffic": true don't seem to be enough.

Neither is adding the VIP ips to the correct nic's allowed_ips field. Adding the VIP's mac. (Which seems to be randomly generated.) manually with dladm to the secondary_macs property doesn't seem to make them work either. 

Allow promisc makes the VIP work, but it has the mentioned unwanted side effects.

This gets really ugly once multiple vlans over a LACP aggr are involved. On ESXi allow promisc is also needed for it to work but it does still seem to do some form of filtering before the traffic hits the vm in that case, as it only gets all stuff untagged from the VLAN the nic is configured for.

**Comment by rmustacc** [comment]
OK, then let's figure out how that normally operates in HA mode and where those packets are getting dropped. Is pfsense using VRRP or CARP or something similar for 'HA mode' or what exactly are those other addresses that are being used?

**Comment by sjorge** [comment]
The *sense appliances seem to use FreeBSD 10 or 11's CARP.

I also run ucarp between a pair of native zones and adding the VIP to the allowed_ips field + mac_spoofing is OK for those.

Is from a working native zone with ucarp.
```
    {
      "interface": "net0",
      "mac": "00:22:06:09:05:01",
      "vlan_id": 5,
      "nic_tag": "external",
      "netmask": "255.255.255.0",
      "ip": "192.168.0.22",
      "ips": [
        "192.168.0.22/24"
      ],
      "allowed_ips": [
        "192.168.0.20/32",
        "224.0.0.18/32"
      ],
      "allow_mac_spoofing": true,
      "mtu": 1500
    },
```

If you have a few scenarios you want me to test with opnsense I can try to do these this weekend.
For now I'm still rebuilding my firewall infrastructure after a crash and a day and a half of head scratching on this issue. (Without HA for now obviously)

**Comment by plitc** [comment]
I do not know exactly how the OpnSense (FreeBSD) Kernel packed up the Proto 112 (CARP + Multicast) and push it over the VirtIO layer (would not be a tap device the right one for the promiscuous mode on kvm side to get all frame types?)

I do not see any special magic behind CARP or uCARP (-M --nomcast / Use broadcast instead of multicast advertisements)

What needs to be fixed so it works without promiscuous mode ...

**Comment by sjorge** [comment]
@plitc do you still have the test setup?
I don't have enough room on one of my compute nodes so I can't build one right now.

I wonder if adding the IP to allowed_ips and manually setting the mac of the CARP interface via dladm in the secondary-macs property might work.

**Comment by plitc** [comment]
@sjorge yes I can build a laboratory environment, copy the two productive opnsense 17.1.5 and test, currently under smartos 20170303

what should I test? different settings?

**Comment by sjorge** [comment]
@plitc I was thinking about the basic VIP setup WAN/LAN + 2x VIP as required.

But instead of setting allow_promisc you lookup the mac of the VIP.
Then add the IP's of the VIPS to allowed_ips via vmadm and add the macs to the proper nics using dladm from the global zone.

```dladm set-linkprop -z UUID-of-zone -p secondary-macs=42:8e:a5:23:0e:5b vnic0```

@rmustacc that should theoretically work right? Adding all the used IP's and MAC's manually?

**Comment by sjorge** [comment]
I bumped into this again, luckily I remember this ticket. I should probably dig a bit deeper into this once I have more time to do so.

**Comment by liv3010m** [comment]
Is this working as expected now?

I remember doing something similar on Solaris 11.3 and Oracle VM for SPARC where creating virtual nics with ldm add-vnet .... alt-mac-addrs=auto|MAC-address[,auto|MAC-address,...] was needed for creating vnics with more than one MAC address. This was for CARP too but on a OpenBSD guest domains/VM.

Thanks,
Olivia

**Comment by sjorge** [comment]
AFAIK this is still broken for KVM and it is also broken for bhyve

**Comment by liv3010m** [comment]
Thanks Jorge for your fast response!
Sad to hear that. So it's still either running ucarp + ipfilter on a zone, running smartos virtualized in parallel to other VMs, or if we can, having multiple physical hosts altogether.

Olivia

**Comment by kilt88** [comment]
I've just discovered this thread, as I'm having the same issue making similar attempts on OpenBSD 6.6 + Bhyve, building a pair of HA firewalls with full connection tracking sync.

SmartOS 20191205T020950Z

I even tried pulling vlans out of the interface within the vm, while using allow_unfiltered_promisc, which didn't work, either.  I'd really like this to work, because I want pfsync that comes with bsd, and unfortunately does not with Illumos.

I'll make a few more attempts with the above idea of adding the secondary mac addresses to the link properties.  But if that doesn't work, I may resort to a dedicated nic passed through to the vm.

**Comment by kilt88** [comment]
I can confirm that adding the secondary mac to linkprop works, though this has to be run outside of vmadm ... I'll likely maintain this with cron, as I've not [yet] found a way to launch scripts based on a zone going up or down.

I have two hardware servers, with each running an LACP aggregate on an Intel i350 nic.  The external nic_tag sits on that aggregate, carrying only tagged vlans to SmartOS.  My two bhyve openbsd firewalls run on these two servers.

I have net0 configured as follows:
  {
    "interface": "net0",
    "mac": "aa:aa:aa:aa:aa:aa",
    "vlan_id": 99,
    "nic_tag": "external",
    "model": "virtio",
    "allow_ip_spoofing": true,
    "allow_mac_spoofing": true,
    "allow_restricted_traffic": true,
    "primary": true
  }

I'm not sure why 'allow_mac_spoofing' isn't solving this for me, however.  But whatever -- the set-linkprop workaround is easy....

Once the vm is started, I issue:
dladm set-linkprop -z $uuid -p secondary-macs=bb:bb:bb:bb:bb:bb net0

Within openbsd, /etc/hostname.carp0:
inet 192.168.99.1 255.255.255.0 192.168.99.255 carpdev vio0 vhid 100 pass bignastypassword advbase 1 advskew 0

You'll notice I did not set 'carppeer' in carp0's config, because multicast communication is working between the two VMs, across the the vlan, across the aggregate to the switch, then to the other vm on the other smartos host ... awesome!

Now to test pfsync....

**Comment by liv3010m** [comment]
Hey, sorry to revive this, but wanted to know if anyone made some progress on it, specially on bhyve since it is making a lot of advancements lately.

**Comment by jasonbking** [comment]
Is the issue now that we just need to add support for secondary MAC addresses (and plumb that up through vmadm)?

**Comment by sjorge** [comment]
@jasonbking secondary MAC's and IPs that are not configured on an interface.

I guess for CARP specifically you need either:
allow_mac_spoofing + allow_ip_spoofing
or
extra mac for CARP tunnel (based on tunnel IP and IPv4/IPv6, sort of like VRRP) and the IP that the tunnel will use...

So you need to white list the ipv4/ipv6 address too... without having it already configured on the vnic (I guess for bhyve that is not an issue, but kvm and native zones will currently always configure it)
